handler/src/main/java/io/netty/handler/pcap/WritePCAPHandler.java

```diff
@@ -0,0 +1,159 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.pcap;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.Inet4Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+
+public final class WritePCAPHandler extends ChannelDuplexHandler {
+
+    private final Protocol protocol;
+    private final PCapFileWriter pCapFileWriter;
+
+    public WritePCAPHandler(Protocol protocol, String destinationFile) throws IOException {
+        this(protocol, new File(destinationFile));
+    }
+
+    public WritePCAPHandler(Protocol protocol, File destinationFile) throws IOException {
+        this.protocol = protocol;
+        this.pCapFileWriter = new PCapFileWriter(destinationFile);
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        writePacket(ctx, msg, false);
+        super.channelRead(ctx, msg);
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        writePacket(ctx, msg, true);
+        super.write(ctx, msg, promise);
+    }
+
+    private void writePacket(ChannelHandlerContext ctx, Object msg, boolean isWrite) throws IOException {
+        if (msg instanceof ByteBuf) {
+            // Duplicate the ByteBuf
+            ByteBuf packet = ((ByteBuf) msg).duplicate();
+
+            InetSocketAddress srcAddr;
+            InetSocketAddress dstAddr;
+            /*
+             * When `isWrite` it true, it means we're sending data from Netty to somewhere else.
+             * In this case, source address will be `localAddress` and destination address will
+             * be `remoteAddress`.
+             *
+             * When `isWrite` is false, it means we're reading data from somewhere else in Netty.
+             * In this case, source address will be `remoteAddress` and destination address will
+             * be `localAddress`.
+             */
+            if (isWrite) {
+                srcAddr = (InetSocketAddress) ctx.channel().localAddress();
+                dstAddr = (InetSocketAddress) ctx.channel().remoteAddress();
+            } else {
+                srcAddr = (InetSocketAddress) ctx.channel().remoteAddress();
+                dstAddr = (InetSocketAddress) ctx.channel().localAddress();
+            }
+
+            if (protocol == Protocol.TCP) {
+                handleTCP(ctx, srcAddr, dstAddr, packet);
+            } else {
+                handleUDP(ctx, srcAddr, dstAddr, packet);
+            }
+        }
```

Comment:
nit: I wonder if you also want to handle `DatagramPacket` for `protocol == Protocol.UDP` ?